### DISCUSSION TITLE
Use withMaxDepth but eager load multiple deeper nested paths

### DISCUSSION THREAD

**Discussion by stebrl** [staudenmeir/laravel-adjacency-list#236] [Q&A]
I have a Livewire app where I want to display a user hierarchy starting from the root user, but only load up to a defined depth:
```
    public function getRootUserProperty(): User
    {
        return User::withMaxDepth(
            $this->depth,
            fn () => User::query()
                ->treeOf(fn ($query) => $query->where('id', $this->root->id))
                ->withCount('children') // to display the children count in the frontend
                ->get()
                ->toTree()
                ->first()
        );
    }
```

All good. Now I want to load a users children (the user I want to expand is always located at the current max depth) and go as deep as I want without modifying the global depth, while keeping the expanded paths in the query string.
Since I have thousands of users and a minimum of 8-10 depth levels, I definitely need to use User::withMaxDepth to keep the query as efficient as possible, right? But how do I expand those specific paths in the existing query? Can you think of a different approach?

Thanks!

**Comment by staudenmeir** [comment]
Hi @stebrl,

> Now I want to load a users children

You want to load the children of every user in this tree, right?

> while keeping the expanded paths in the query string.

How do you mean that? Can you share an example?

**Comment by stebrl** [comment]
Let´s say I have a tree loaded with a maximum depth of 2 that looks like:
```
Root
    User1
        User11
        User12
        User13
    User2
        User21
    User3
        User31
        User32
```
like the example code I provided earlier.

Now I want to expand User12 to only see this users direct children (not all descendants), so the tree looks like:
```
Root
    User1
        User11
        User12
            User121
            User122
            User123
        User13
    User2
        User21
    User3
        User31
        User32
```
This information should be synced in the query string, so I can share the link or do whatever I want with it.

Now I also want to expand User121s children so the tree looks like:
```
Root
    User1
        User11
        User12
            User121
                User1211
                User1212
            User122
            User123
        User13
    User2
        User21
    User3
        User31
        User32
```
It´s tricky because we also need to load User121s siblings + User121s children to recreate this tree. Also I honestly don´t know what to store in the query string in this case. Maybe the expanded path (12.121 in this case). All of this should also work for even deeper children.
I hope you get the idea now, sorry for confusion!

**Comment by staudenmeir** [comment]
> This information should be synced in the query string, so I can share the link or do whatever I want with it.

Can you describe the overall context here? Does every user have its own page? Is the root user always the same?

> Now I also want to expand User121s children so the tree looks like:

When does this expansion need to happen?

**Reply by stebrl** [reply]
> Can you describe the overall context here? Does every user have its own page? Is the root user always the same?

My app has admins and normal users. Admins do stuff with users and see the same tree originating from the same root user. It´s like an organization hierarchy.
Users do not have their own tree yet.

> When does this expansion need to happen?

When an admin is "clicking through" the tree, expanding multiple nodes to see children, grand children etc.
I want to save the tree´s state in the query string so when they reload the page, everything is expanded like before.

**Reply by staudenmeir** [reply]
I see. A few ideas:
- You could store the IDs of the expanded users (the ones the user clicked on) in the URL. Also add a second parameter that contains the maximum depth the user "added" (click on `User12` -> 1, click on `User13` -> still 1, click on `User121` -> 2).
- When the page is loaded with query parameters, you load the tree with a depth of `$this->depth + $maxAddedDepth`.
- Before applying `->toTree()`, you remove the users with a depth `> $this->depth` that don't have an ancestor in the list of expanded users (check their `path`).
- If the list of expanded IDs gets too long, you could only store IDs of the deepest users (e.g. `121` instead of `12` & `121`)
